src/txn/txn_rollback_to_stable.c

```diff
@@ -1224,54 +1224,40 @@ __rollback_abort_updates(WT_SESSION_IMPL *session, WT_REF *ref, wt_timestamp_t r
 }
 
 /*
- * __rollback_abort_fast_truncate --
- *     Abort fast truncate for an internal page of leaf pages.
+ * __rollback_to_stable_page_skip --
+ *     Skip if rollback to stable doesn't require reading this page.
  */
 static int
-__rollback_abort_fast_truncate(
-  WT_SESSION_IMPL *session, WT_REF *ref, wt_timestamp_t rollback_timestamp)
-{
-    WT_REF *child_ref;
-
-    WT_INTL_FOREACH_BEGIN (session, ref->page, child_ref) {
-        /*
-         * A fast-truncate page is either in the WT_REF_DELETED state (where the WT_PAGE_DELETED
-         * structure has the timestamp information), or in an in-memory state where it started as a
-         * fast-truncate page which was then instantiated and the timestamp information moved to the
-         * individual WT_UPDATE structures. When reviewing internal pages, ignore the second case,
-         * an instantiated page is handled when the leaf page is visited.
-         */
-        if (child_ref->state == WT_REF_DELETED && child_ref->ft_info.del != NULL &&
-          rollback_timestamp < child_ref->ft_info.del->durable_timestamp) {
-            __wt_verbose_multi(session, WT_VERB_RECOVERY_RTS(session),
-              "%p: deleted page rolled back", (void *)child_ref);
-            WT_RET(__wt_delete_page_rollback(session, child_ref));
-        }
-    }
-    WT_INTL_FOREACH_END;
-    return (0);
-}
-
-/*
- * __wt_rts_page_skip --
- *     Skip if rollback to stable doesn't requires to read this page.
- */
-int
-__wt_rts_page_skip(
+__rollback_to_stable_page_skip(
   WT_SESSION_IMPL *session, WT_REF *ref, void *context, bool visible_all, bool *skipp)
 {
+    WT_PAGE_DELETED *page_del;
     wt_timestamp_t rollback_timestamp;
 
-    rollback_timestamp = *(wt_timestamp_t *)(context);
+    rollback_timestamp = *(wt_timestamp_t *)context;
+    WT_UNUSED(visible_all);
+
     *skipp = false; /* Default to reading */
 
-    WT_UNUSED(visible_all);
+    /*
+     * Skip unprepared or resolved fast-truncate operations durable at or before the RTS timestamp.
+     * If we find a page without fast-truncate timestamp information, it's an old format page:
+     * there's no way to get correct behavior, skip them, that's what old versions of this code did.
+     */
+    if (ref->state == WT_REF_DELETED) {
+        page_del = ref->ft_info.del;
+        if (page_del == NULL ||
+          ((page_del->prepare_state == 0 || page_del->prepare_state == WT_PREPARE_RESOLVED) &&
```

Comment:
Do we need to check for the checkpoint snapshot also here? In case the proxy cell is written by an eviction thread that has a snapshot greater than the checkpoint.

Reply:
yes, agree

Reply:
Fixed in 053e357.

Would you please review this change again? -- Thanks.